### DISCUSSION TITLE
How to save all test data images

### DISCUSSION THREAD

**Discussion by bibinwils** [Lightning-AI/pytorch-lightning#9226] [code help: CV]
Hi, I trained a UNet with input images and mask. For that I split the entire dataset to train - valid - test. I have trained the network and got the output metrics for the test dataset. There are around 100 test images. I would like to pass through the model and save the corresponding output images to a folder. How to save the test output so?

**Comment by rohitgr7** [comment]
for that you can try `trainer.predict` by passing the test dataloader inside predict. https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#predict

**Reply by bibinwils** [reply]
Hi, Thanks for the reply. May I know how do I save all the images?

**Reply by rohitgr7** [reply]
https://discuss.pytorch.org/t/how-can-i-save-generate-image-in-new-separate-folder/84080/2
https://discuss.pytorch.org/t/converting-tensors-to-images/99482/6
https://discuss.pytorch.org/t/how-to-save-image/48808

**Comment by bibinwils** [comment]
Hi, I couldn't succeed. Can you help me out? I tried like
Trainer.predict(model=clf)
for i in range(100):
torchvision.utils.save_image(image_list[-1][i], “/images/generated_image%d.JPG” %i)

but not getting any results

**Reply by bibinwils** [reply]
anybody here to help?

**Comment by tchaton** [comment]
Won't this work ?

```py
predictions = trainer.predict(model=clf)
for i in range(100):
    torchvision.utils.save_image(predictions[-1][i], “/images/generated_image%d.JPG” %i)
```

**Reply by bibinwils** [reply]
no, it's not working

**Reply by bibinwils** [reply]
MisconfigurationException: Dataloader not found for `Trainer.predict` - it's the error coming there

**Reply by bibinwils** [reply]
import torchvision
predictions = trainer.predict(model=clf)

for i in range(32):

    torchvision.utils.save_image(predictions[-1][i], 'generated_image'+'%d'+'.JPG' %i)



This is what I tried. I put 32 since my batch size is 32

**Reply by rohitgr7** [reply]
did you define a predict_dataloader? if not you need to pass one inside `trainer.predict`.

**Reply by bibinwils** [reply]
Thanks. I tried. Now i'm getting this error
![image](https://user-images.githubusercontent.com/53151992/132549824-e25cbd8a-8724-4afd-8ef9-e053677b4edc.png)

**Reply by rohitgr7** [reply]
I don't remember what the exact dimension of the predictions is supposed to be, but you can analyze the prediction output a little bit and figure out how to access the required image output. If it doesn't work out pls share a minimal example, I'll check it :)
